index.js

```diff
@@ -112,6 +122,43 @@ class Conf {
 		}
 	}
 
+	_migrate(migrations = {}, versionToMigrate = '0.0.0') {
+		const MIGRATION_KEY = '__conf-migrated-version__';
+
+		const previousMigratedVersion = this.get(MIGRATION_KEY, '0.0.0');
+
+		const olderVersionsWithMigrations = Object.keys(migrations)
+			.filter(version => this._shouldVersionPerformMigration(version, previousMigratedVersion, versionToMigrate))
+			.sort(semver.compare);
+
+		const migrationsToRun = olderVersionsWithMigrations.map(version => migrations[version]);
+
+		migrationsToRun.forEach(migration => migration(this));
```

Comment:
What happens if one of the migration functions fail? Should we have to worry about this by having some sort of "restore snapshot on error"? Or should we simply say that the migration functions can run more than once for the same version, meaning they must be idempotent?

Reply:
I'm more inclined to the second.

Reply:
Nice catch!

So what you're saying in the first example is:

```
1.0.0 -> success (set migration_key to 1.0.0)
1.0.1 -> success (set migration_key to 1.0.1)
1.0.2 -> fails (throw error, leave migration_key on 1.0.1 and restore previous changes to the store)
```

and in the second


```
1.0.0 -> success 
1.0.1 -> success 
1.0.2 -> fails (throw error, leave migration_key on 0.0.0 and restore previous changes to the store)
```

?

Reply:
You are almost correct, though in the second example:

- The migration_key would be updated to 1.0.1.
- No changes would be restored. The handler for `1.0.2` would be run again in the future, meaning that it should be idempotent. By idempotent, I mean ensuring that the outcome of the overall migration for that version is the same, regardless of a previous partly successful run.

The second approach is much easier to implement. The first example is difficult to implement because of the "restore previous changes". What happens if we fail to restore the previous changes too? Since we are not dealing with a store that supports transactions, this would be almost impossible to implement correctly on failures.

Reply:
Storing the values before the mutations could be possible by using the `onDidChange` and using the `callback` to temporarily store the store state.

I do agree that your suggestion is simpler and works fine (sorry for miss-understanding your text), so I don't think there's a need to over-complicate with the "restore state". 

I will add the code to allow for this and also update the documentation in both the repositories. Thank you for clearing that up!

Reply:
Done, please review again when you have time 😄

Reply:
Done!